src/ncp/hdlc.hpp

```diff
@@ -259,83 +260,77 @@ template <uint16_t kSize> class MultiFrameBuffer : public FrameWritePointer
     {
-        if (!HasSavedFrame())
-        {
-            Clear();
-        }
-        else
-        {
-            mWritePointer    = mWriteFrameStart + kHeaderSize;
-            mRemainingLength = static_cast<uint16_t>(mBuffer + kSize - mWritePointer);
-        }
+        mWritePointer    = mWriteFrameStart + kHeaderSize;
+        mRemainingLength = static_cast<uint16_t>(mBuffer + kSize - mWritePointer);
     }
 
     /**
-     * This method indicates whether there are unread saved frames in the buffer.
+     * This method indicates whether there are any saved frames in the buffer.
      *
-     * @retval TRUE  There is at least one unread saved frame in buffer.
-     * @retval FALSE There is no unread saved frame in buffer.
+     * @retval TRUE  There is at least one saved frame in the buffer.
+     * @retval FALSE There is no saved frame in the buffer.
      *
      */
-    bool HasSavedFrame(void) const { return (mSavedFrameStart != mWriteFrameStart); }
+    bool HasSavedFrame(void) const { return (mWriteFrameStart != mBuffer); }
 
     /**
-     * This method reads a previously saved frame from the buffer.
-     *
-     * Subsequent call to this method reads the next saved frame in the buffer (if any).
+     * This method iterates through previously saved frames in the buffer, getting a next frame in the queue.
      *
-     * @param[out] aFrame   A reference to a pointer variable to return start of the frame.
-     * @param[out] aLength  A reference to a variable to return the frame length (number of bytes).
+     * @param[inout] aFrame   On entry, should point to a previous saved frame or NULL to get the first frame.
+     *                        On exit, the pointer variable is updated to next frame or set to NULL if there are none.
+     * @param[out]   aLength  A reference to a variable to return the frame length (number of bytes).
      *
-     * @retval OT_ERROR_NONE       Frame was read successfully, @p aFrame and @p aLength updated.
-     * @retval OT_ERROR_NOT_FOUND  No more saved frame in buffer.
+     * @retval OT_ERROR_NONE       Updated @aFrame and @aLength successfully with the next saved frame.
+     * @retval OT_ERROR_NOT_FOUND  No more saved frame in the buffer.
      *
      */
-    otError ReadSavedFrame(uint8_t *&aFrame, uint16_t &aLength)
+    otError GetNextSavedFrame(uint8_t *&aFrame, uint16_t &aLength)
     {
         otError error = OT_ERROR_NONE;
 
-        if (mSavedFrameStart != mWriteFrameStart)
+        assert(aFrame == NULL || (mBuffer <= aFrame && aFrame < OT_ARRAY_END(mBuffer)));
+
+        aFrame = (aFrame == NULL) ? mBuffer : aFrame + Encoding::LittleEndian::ReadUint16(aFrame - kHeaderSize);
+
+        if (aFrame != mWriteFrameStart)
         {
-            aLength = Encoding::LittleEndian::ReadUint16(mSavedFrameStart);
-            aFrame  = mSavedFrameStart + kHeaderSize;
-            mSavedFrameStart += aLength + kHeaderSize;
+            aLength = Encoding::LittleEndian::ReadUint16(aFrame);
```

Comment:
Maybe we don't need deal with byte order because this is not transmitted in network.

Reply:
Agree. Let's keep as is since won't hurt to be specific and we have the nice `ReadUint16()` method. :)